File: src/io/innofang/java/string/KMP.java
package io.innofang.java.string;

/**
 * Created by Inno Fang on 2018/3/15.
 */
public class KMP {

    private char[] pattern;
    private int[] next;

    public KMP(char[] pattern) {
        resetPattern(pattern);
    }

    public KMP(String pattern) {
        resetPattern(pattern);
    }

    public void resetPattern(String pattern) {
        this.pattern = pattern.toCharArray();
        next = generateNext(pattern.length());
    }

    public void resetPattern(char[] pattern) {
        this.pattern = pattern;
        next = generateNext(pattern.length);
    }
    public boolean matches(String text) {
        return matches(text.toCharArray());
    }

    public boolean matches(char[] text) {
        int t = 0;
        int p = 0;
        int n = 0;
        while (t < text.length && p < pattern.length) {
            if (text[t] == pattern[p]) {
                t++;
                p++;
            } else if (p != 0) p = next[p - 1];
            else t++;
        }
        return p == pattern.length;
    }

    private int[] generateNext(int length) {
        int[] tmp = new int[length];
        int idx = 0;
        tmp[idx] = 0;
        for (int i = 1; i < length; ) {
            if (pattern[idx] == pattern[i]) tmp[i++] = (idx++) + 1;
                // tmp[idx] != tmp[i]
            else if (idx == 0) tmp[i++] = 0;
            else idx = tmp[idx - 1];
        }
        return tmp;
    }

}

File: src/io/innofang/kotlin/string/KMP.kt
package io.innofang.kotlin.string

/**
 * Created by Inno Fang on 2018/3/16.
 */
class KMP
constructor(private var pattern: CharArray) {

    private lateinit var next: IntArray

    init {
        resetPattern(pattern)
    }


    constructor(pattern: String) : this(pattern.toCharArray()) {
        resetPattern(pattern)
    }

    fun resetPattern(pattern: String) = resetPattern(pattern.toCharArray())


    fun resetPattern(pattern: CharArray) {
        this.pattern = pattern
        next = generateNext(pattern.size)
    }

    fun matches(text: String): Boolean = matches(text.toCharArray())


    fun matches(text: CharArray): Boolean = find(text) >= 0


    fun find(text: String): Int = find(text.toCharArray())


    fun find(text: CharArray): Int {
        var t = 0
        var p = 0
        while (t < text.size && p < pattern.size) {
            when {
                text[t] == pattern[p] -> {
                    t++
                    p++
                }
                p != 0 -> p = next[p - 1]
                else -> t++
            }
        }
        return if (p == pattern.size) t - p else -1
    }

    private fun generateNext(length: Int): IntArray {
        val tmp = IntArray(length)
        var idx = 0
        tmp[idx] = 0
        var i = 1
        while (i < length) {
            when {
                pattern[idx] == pattern[i] -> tmp[i++] = idx++ + 1
                idx == 0 -> tmp[i++] = 0
                else -> idx = tmp[idx - 1]
            }
        }
        return tmp
    }
}
